src/lib/ERC4337/account/Account.sol

```diff
@@ -0,0 +1,190 @@
+import {IERC1271} from "openzeppelin-contracts/interfaces/IERC1271.sol";
+
+/// @title Station Network Account Abstract Contract
+/// @author 👦🏻👦🏻.eth
+
+/// @dev This abstract contract provides scaffolding for Station's Account signature validation
+/// ERC1271 and ERC4337 compliance in combination with Mage's Permissions system
+/// to provide convenient and modular private key management on an infrastructural level
+abstract contract Account is Mage, IAccount, IERC1271, ModularValidationInternal {
+
+    /*=============
+        ACCOUNTS
+    ==============*/
+    
+    /// @dev This chain's EntryPoint contract address
+    address public immutable entryPoint;
+
+    /// @param _entryPointAddress The contract address for this chain's ERC-4337 EntryPoint contract
+    /// Official address for the most recent EntryPoint version is `0x5FF137D4b0FDCD49DcA30c7CF57E578a026d2789`
+    constructor(address _entryPointAddress) {
+        entryPoint = _entryPointAddress;
+    }
+
+    /// @dev Function enabling EIP-4337 compliance as a smart contract wallet account
+    /// @param userOp The UserOperation to validate before executing
+    /// @param userOpHash Hash of the UserOperation data, used as signature digest
+    /// @param missingAccountFunds Delta representing this account's missing funds in the EntryPoint contract
+    /// Corresponds to minimum native currency that must be transferred to the EntryPoint to complete execution
+    /// Can be 0 if this account has already deposited enough funds or if a paymaster is used
+    /** 
+    *   @return validationData A packed uint256 of three concatenated variables
+    *   ie: `uint256(abi.encodePacked(address authorizor, uint48 validUntil, uint48 validAfter))`
+    *   where `authorizer` can be one of the following:
+    *       1. A signature aggregator contract, inheriting IAggregator.sol, to use for validation
+    *       2. An exit status code `bytes20(0x01)` representing signature validation failure 
+    *       3. An empty `bytes20(0x0)` representing successful signature validation
+    */
+    function validateUserOp(
+        UserOperation calldata userOp, 
+        bytes32 userOpHash, 
+        uint256 missingAccountFunds
+    ) public virtual returns (uint256 validationData) {
+        // only EntryPoint should call this function to prevent frontrunning of valid signatures
+        _checkSenderIsEntryPoint();
+
+        // extract validator address using cheap calldata slicing before decoding 
+        address validator = address(bytes20(userOp.signature[12:32]));
+
+        if (isValidator(validator)) {
+            ( , address signer, bytes memory nestedSig) = abi.decode(userOp.signature, (address, address, bytes));
+            // copy userOp into memory and format for Validator module
+            UserOperation memory formattedUserOp = userOp;
+            formattedUserOp.signature = abi.encode(signer, nestedSig);
```

Comment:
I spent a while mulling over this as well, let's sync on it.

The context for this was to maintain consistency between `isValidSignature()` and `validateUserOp()` in the face of a Solidity limitation:

- The EIP1271 spec requires `isValidSignature()` to take its `signature` parameter as a bytes array _in memory_ whereas the 4337 spec defines its respective `signature` parameter as a bytes array within the `userOp` struct located _in calldata_
- Solidity unfortunately doesn't offer byte array slicing when operating in memory, so I was left with four options:

To extract the `validator`, `signer`, and `nestedSignature`, either:
1. Use calldata slicing for `validateUserOp` and use redundant abi decode + re-encode for `isValidSignature()`
2. Use assembly for one or both functions
3. Use the redundant abi decode and re-encode for both functions 

I went with option 3 so the functions would look more symmetrical and maybe be more readable but I am of course not married to the approach and happy to change it.

Reply:
In removing the `signer` parameter and implementing a check for `VALIDATOR_FLAG` + `address(validator)` it became clear that using abi decoding can revert in some cases when the data is not formatted to expectations.

As a result, `validateUserOp()` uses calldata slicing since it is available and `isValidSignature()` uses one line of assembly to accomplish the same task but in memory.